torchqdynamics/utils/td_tensor.py

```diff
@@ -0,0 +1,172 @@
+        'Time-dependent operators in the `callable` format should always'
+        ' return a `torch.Tensor` with the same dtype and device as provided'
+        ' to the solver. This avoids type conversion or device transfer at'
+        ' every time step that would slow down the solver.'
+    )
+    if not isinstance(tensor, Tensor):
+        raise TypeError(
+            f'{prefix} The provided operator is currently of type'
+            f' {type(tensor)} instead of {Tensor}.'
+        )
+    elif tensor.dtype != expected_dtype:
+        raise TypeError(
+            f'{prefix} The provided operator is currently of dtype'
+            f' {tensor.dtype} instead of {expected_dtype}.'
+        )
+    elif tensor.device != expected_device:
+        raise TypeError(
+            f'{prefix} The provided operator is currently on device'
+            f' {tensor.device} instead of {expected_device}.'
+        )
+
+    return tensor
+
+
+class TDTensor(ABC):
+    @abstractmethod
+    def __call__(self, t: float) -> Tensor:
+        """Evaluate at a given time"""
+        pass
+
+    @abstractmethod
+    def size(self, dim: int) -> int:
+        """Size along a given dimension."""
+        pass
+
+    @abstractmethod
+    def dim(self) -> int:
+        """Get the number of dimensions."""
+        pass
+
+    @abstractmethod
+    def unsqueeze(self, dim: int) -> TDTensor:
+        """Unsqueeze at position `dim`."""
+        pass
+
+    @abstractmethod
+    def requires_updates(self, t: float) -> bool:
+        """Whether the cache needs to be updated for new time values."""
+        pass
+
+
+class ConstantTDTensor(TDTensor):
+    def __init__(self, tensor: Tensor):
+        self._tensor = tensor
+        self.dtype = tensor.dtype
+        self.device = tensor.device
+
+    def __call__(self, t: float) -> Tensor:
+        return self._tensor
+
+    def size(self, dim: int) -> int:
+        return self._tensor.size(dim)
+
+    def dim(self) -> int:
+        return self._tensor.dim()
+
+    def unsqueeze(self, dim: int) -> ConstantTDTensor:
+        return ConstantTDTensor(self._tensor.unsqueeze(dim))
+
+    def requires_updates(self, t: float) -> bool:
+        return False
+
+
+class CallableTDTensor(TDTensor):
+    def __init__(
+        self,
+        f: callable,
+        *,
+        shape: torch.Size | None = None,
+        dtype: torch.dtype | None = None,
+        device: torch.device | None = None,
+    ):
+        self._callable = f
+        self.dtype = dtype
+        self.device = device
+        self._shape = shape
+        self._cached_tensor = None
+        self._cached_t = None
+
+    def __call__(self, t: float) -> Tensor:
+        if t != self._cached_t:
+            self._cached_tensor = self._callable(t).view(self._shape)
+            self._cached_t = t
+        return self._cached_tensor
+
+    def size(self, dim: int) -> int:
+        return self._shape[dim]
+
+    def dim(self) -> int:
+        return len(self._shape)
+
```

Comment:
If `shape` is not given in the constructor, this will always fail. Maybe it's better to update `self.shape` in `__call__` and then in `size` and `dim` if `shape` is not set then we use `__call__` to determine the shape

Reply:
Well the way I was thinking about the class was that `shape` was kinda mandatory. Would it be fine if I just remove the default argument of `shape` in the `__init__`, and remove the typing `None` ?